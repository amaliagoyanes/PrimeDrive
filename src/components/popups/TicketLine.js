import React, {Component} from 'react';
import {
    Dimensions,
    TouchableOpacity,
    Text,
    View,
    TextInput,
} from 'react-native';
import {FormattedMessage} from 'react-intl';
import {Row, Col} from 'react-native-easy-grid';
import styles from '../../style/styles';
import Modal from 'react-native-root-modal';
import {NumericKeyboard} from '../index.js';
import {scale, scaleText} from '../../scaling';
import {toFixedLocale} from '../plain/receipts/plainComponents';


export default class TicketLine extends Component {
    constructor(props) {
        super(props);
        this.state = {
            input: '', active: 0, open: false, px: 0, py: 0, uid: null, name: ''
        };

        this.setActive = this.setActive.bind(this);
        this.confirm = this.confirm.bind(this);
        this.close = this.close.bind(this);
        this.open = this.open.bind(this);
    }

    setActive(i) {
        this.setState({active: i})
    }

    open(px, py, amount = 1, name = '', uid = null, note = '', product_price = 0, discount = '', campaign = '') {

        this.setState({open: true, px: px, py: py, name: name, amount: amount, uid: uid, note: note,
                       product_price: product_price, discount: discount, campaign: campaign})
    }

    close() {
        this.setState({open: false})
    }

    confirm() {
        let values = this.controls.getVal();
        if (this.props.onOk !== undefined) {
            this.props.onOk(this.amountCtrl.getVal(), this.state.uid, undefined, values.note, values.discount);
        }
        this.close();
    }

    render() {
        let left = this.props.arrow === 'left' ?
            this.state.px + 10 : this.state.px - Dimensions.get('screen').width / 6 - 10;
        let leftArrow = this.props.arrow === 'left' ?
            this.state.px : this.state.px - 15;
        let isDiscount = this.state.active === 0 && !this.props.readOnly;
        let isCampaign = this.state.campaign && this.state.campaign.amount;
        let campaign_amount = '';
        let campaign_text = '';
        if (isCampaign && this.state.campaign.type === 'discount') {
            campaign_amount = toFixedLocale(this.state.product_price * this.state.amount * (parseFloat(this.state.campaign.amount.replace(',', '.')) / 100));
            campaign_text = `${this.props.intl.messages["app.campaign"]} ${this.state.campaign.name} - ${this.state.campaign.amount}%`;
        }
        if (isCampaign && this.state.campaign.type === 'price') {
            campaign_amount = toFixedLocale(this.state.campaign.amount.replace(',', '.') * this.state.amount);
            campaign_text = `${this.props.intl.messages["app.campaign"]} ${this.state.campaign.name} - ${this.props.intl.messages["app.price"]} ${this.state.campaign.amount} DKK`;
            // style.textDecorationLine = 'line-through';
        }

        return (
            <Modal
                onRequestClose={this.close}
                visible={this.state.open}
                style={styles.modal}
            >
                <TouchableOpacity style={styles.overlay} onPress={this.close}>
                    <View style={{
                        position: 'absolute',
                        top: this.state.py + 5,
                        left: leftArrow
                    }}>
                        <View
                            style={this.props.arrow === 'left' ? [styles.triangle, styles.leftTriangle] : styles.triangle}/>
                    </View>
                    <TouchableOpacity activeOpacity={1} onPress={(event) => event.stopPropagation()} style={{
                        position: 'absolute',
                        top: this.state.py - 20,
                        left: left,
                        width: Dimensions.get('screen').width / 6,
                        height: isDiscount ? 300 : 155,
                        backgroundColor: '#404048'
                    }}>
                        <Row size={!!campaign_amount ? 1 : 2}
                            style={{alignItems: 'center', justifyContent: 'center', padding: scale(5)}}>
                            <Text style={styles.text}>{this.state.name.toUpperCase()}</Text>
                        </Row>
                        {!!campaign_amount &&
                        <Row style={{alignItems: 'center', justifyContent: 'center', marginLeft: scale(5), marginRight: scale(5)}}>
                            <Text style={[styles.text, styles.sectionDiscount]}>{campaign_text}</Text>
                            <Text style={[styles.text, styles.sectionDiscountPrice]}>{this.state.campaign.type==='discount'?'-':''}{campaign_amount} DKK</Text>
                        </Row>
                        }

                        <AmountControl ref={elt => this.amountCtrl = elt} amount={this.state.amount} />
                        <Row style={{backgroundColor: '#6f7076', marginTop: 1}}>
                            <TouchableOpacity
                                style={this.state.active === 0 ? [styles.tabButton, styles.activeColor] : [styles.tabButton]}
                                onPress={() => this.setState({active: 0})}>
                                <Text style={[styles.text, {fontSize: scaleText(9)}]}>
                                    {this.props.intl.messages["app.discountUpper"]}</Text>
                            </TouchableOpacity>
                            <TouchableOpacity
                                style={this.state.active === 1 ? [styles.tabButton, styles.activeColor] : [styles.tabButton]}
                                onPress={() => this.setState({active: 1})}>
                                <Text style={[styles.text, {fontSize: scaleText(9)}]}>
                                    {this.props.intl.messages["app.noteUpper"]}</Text>
                            </TouchableOpacity>
                            <TouchableOpacity style={[styles.tabButton, {backgroundColor: '#2eab61'}]}
                                              onPress={this.confirm}>
                                <Text style={[styles.text, {fontSize: scaleText(9)}]}>
                                    {this.props.intl.messages["app.okUpper"]}
                                </Text>
                            </TouchableOpacity>
                        </Row>
                        <TicketLineControls active={this.state.active} isDiscount={isDiscount} intl={this.props.intl}
                                            discount={this.state.discount} note={this.state.note} alert={this.props.alert}
                                            ref={elt => this.controls = elt} />
                    </TouchableOpacity>
                </TouchableOpacity>
            </Modal>
        )
    }
}

export class TicketLineControls extends Component {
    constructor(props) {
        super(props);
        this.state = {
            input: props.discount || '', text: props.note || ''
        };

        this.numOnPress = this.numOnPress.bind(this);
        this.cleanInput = this.cleanInput.bind(this);
    }

    numOnPress(text) {
        let newVal = this.state.input + text,
            max = 100;
        console.log('this.props.alert ', this.props);
        if (parseFloat(newVal) > max)
            this.props.alert(`${this.props.intl.messages["app.notGreaterThan"]} ${max}`);
        else
            this.setState({input: newVal})
    }

    cleanInput() {
        this.setState({input: ''});
    }

    getVal() {
        return {note: this.state.text, discount: this.state.input};
    }

    render() {
        return (
            <Row size={this.props.isDiscount ? 8 : 2} style={{backgroundColor: '#6f7076', padding: 0}}>
                {this.props.active !== 0 ?
                    <View style={[styles.tabInfo, {alignItems: 'center'}]}>
                        <TextInput value={this.state.text}
                                   style={[styles.text, {fontSize: scaleText(11), width: '100%', height: '100%'}]}
                                   multiline={true}
                                   onChangeText={(text) => this.setState({text})}/>
                    </View> :
                    this.props.readOnly ?
                        <View style={[styles.tabInfo, {alignItems: 'flex-end', padding: 10}]}>
                            <Text style={[styles.text, {fontSize: scaleText(13)}]}>25%</Text>
                        </View> :
                        <Col>
                            <Row size={5}>
                                <NumericKeyboard size={6} numOnPress={this.numOnPress} intl={this.props.intl}
                                                 input={this.state.input} inputBackgroundColor='#6f7076'
                                                 cleanInput={this.cleanInput} smallBlock
                                                 showDelimiter={true} postfix={'%'}/>
                            </Row>
                        </Col>
                }
            </Row>
        )
    }
}

export class AmountControl extends Component {
    constructor(props) {
        super(props);
        this.state = {
            amount: props.amount || 1,
            maxAmount: props.maxAmount

        };

        this.getVal = this.getVal.bind(this);
        this.onPlusPress = this.onPlusPress.bind(this);
    }

    getVal() {
        return this.state.amount;
    }

    onPlusPress() {
        if (this.state.maxAmount) {
            if (this.state.maxAmount > this.state.amount)
                this.setState({amount: this.state.amount + 1});
        }
        else
            this.setState({amount: this.state.amount + 1})
    }

    render() {
        return (
            <Row style={{alignItems: 'center'}}>
                <TouchableOpacity name="minus" style={[styles.tabButton]}
                                  onPress={() => this.state.amount !== 1 ? this.setState({amount: this.state.amount - 1}) : ''}>
                    <Text style={[styles.text]}>-</Text>
                </TouchableOpacity>
                <TouchableOpacity style={[styles.tabButton]}>
                    <Text style={styles.text}>{this.state.amount}</Text>
                </TouchableOpacity>
                <TouchableOpacity name="plus" style={[styles.tabButton]}
                                  onPress={this.onPlusPress}>
                    <Text style={[styles.text]}>+</Text>
                </TouchableOpacity>
            </Row>
        )
    }
}