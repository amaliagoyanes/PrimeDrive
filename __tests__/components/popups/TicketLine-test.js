'use strict';

import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import TicketLine from '../../../src/components/popups/TicketLine';


it('renders correctly', () => {
    const tree = shallow(
        <TicketLine />);
    tree.setState({open: true});
    expect(tree).toMatchSnapshot();
    expect(tree.find('AmountControl').length).toEqual(1);
    expect(tree.find('TicketLineControls').length).toEqual(1);
});

it('renders correctly with left arrow', () => {
    const tree = shallow(
        <TicketLine arrow="left"/>);
    tree.setState({open: true});
    expect(tree).toMatchSnapshot();
});

it('renders correctly with right arrow', () => {
    const tree = shallow(
        <TicketLine arrow="right"/>);
    tree.setState({open: true});
    expect(tree).toMatchSnapshot();
});

