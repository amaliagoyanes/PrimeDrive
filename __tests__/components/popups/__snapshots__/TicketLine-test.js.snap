// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  onRequestClose={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.2)",
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
  visible={true}
>
  <TouchableOpacity
    activeOpacity={0.2}
    focusedOpacity={0.7}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(255,255,255,0.1)",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "left": -15,
          "position": "absolute",
          "top": 5,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "borderBottomColor": "transparent",
            "borderBottomWidth": 10,
            "borderLeftColor": "#404048",
            "borderLeftWidth": 10,
            "borderRightColor": "transparent",
            "borderRightWidth": 0,
            "borderStyle": "solid",
            "borderTopColor": "transparent",
            "borderTopWidth": 10,
            "height": 0,
            "width": 0,
          }
        }
      />
    </View>
    <TouchableOpacity
      activeOpacity={1}
      focusedOpacity={0.7}
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#404048",
          "height": 300,
          "left": -135,
          "position": "absolute",
          "top": -20,
          "width": 125,
        }
      }
    >
      <RowNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Gotham-Book",
              "fontSize": 14,
            }
          }
        >
          CROSSAINT
        </Text>
      </RowNB>
      <AmountControl />
      <RowNB>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
              Object {
                "backgroundColor": "#6f7076",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            DISCOUNT
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            NOTE
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
              Object {
                "backgroundColor": "#2eab61",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            OK
          </Text>
        </TouchableOpacity>
      </RowNB>
      <TicketLineControls
        active={0}
        isDiscount={true}
      />
    </TouchableOpacity>
  </TouchableOpacity>
</Modal>
`;

exports[`renders correctly with left arrow 1`] = `
<Modal
  onRequestClose={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.2)",
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
  visible={true}
>
  <TouchableOpacity
    activeOpacity={0.2}
    focusedOpacity={0.7}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(255,255,255,0.1)",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "top": 5,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "transparent",
              "borderBottomWidth": 10,
              "borderLeftColor": "#404048",
              "borderLeftWidth": 10,
              "borderRightColor": "transparent",
              "borderRightWidth": 0,
              "borderStyle": "solid",
              "borderTopColor": "transparent",
              "borderTopWidth": 10,
              "height": 0,
              "width": 0,
            },
            Object {
              "borderLeftColor": "transparent",
              "borderLeftWidth": 0,
              "borderRightColor": "#404048",
              "borderRightWidth": 10,
            },
          ]
        }
      />
    </View>
    <TouchableOpacity
      activeOpacity={1}
      focusedOpacity={0.7}
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#404048",
          "height": 300,
          "left": 10,
          "position": "absolute",
          "top": -20,
          "width": 125,
        }
      }
    >
      <RowNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Gotham-Book",
              "fontSize": 14,
            }
          }
        >
          CROSSAINT
        </Text>
      </RowNB>
      <AmountControl />
      <RowNB>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
              Object {
                "backgroundColor": "#6f7076",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            DISCOUNT
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            NOTE
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
              Object {
                "backgroundColor": "#2eab61",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            OK
          </Text>
        </TouchableOpacity>
      </RowNB>
      <TicketLineControls
        active={0}
        isDiscount={true}
      />
    </TouchableOpacity>
  </TouchableOpacity>
</Modal>
`;

exports[`renders correctly with right arrow 1`] = `
<Modal
  onRequestClose={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.2)",
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
  visible={true}
>
  <TouchableOpacity
    activeOpacity={0.2}
    focusedOpacity={0.7}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(255,255,255,0.1)",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "left": -15,
          "position": "absolute",
          "top": 5,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "borderBottomColor": "transparent",
            "borderBottomWidth": 10,
            "borderLeftColor": "#404048",
            "borderLeftWidth": 10,
            "borderRightColor": "transparent",
            "borderRightWidth": 0,
            "borderStyle": "solid",
            "borderTopColor": "transparent",
            "borderTopWidth": 10,
            "height": 0,
            "width": 0,
          }
        }
      />
    </View>
    <TouchableOpacity
      activeOpacity={1}
      focusedOpacity={0.7}
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#404048",
          "height": 300,
          "left": -135,
          "position": "absolute",
          "top": -20,
          "width": 125,
        }
      }
    >
      <RowNB
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "fontFamily": "Gotham-Book",
              "fontSize": 14,
            }
          }
        >
          CROSSAINT
        </Text>
      </RowNB>
      <AmountControl />
      <RowNB>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
              Object {
                "backgroundColor": "#6f7076",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            DISCOUNT
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            NOTE
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#404048",
                "justifyContent": "center",
                "width": "34%",
              },
              Object {
                "backgroundColor": "#2eab61",
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontFamily": "Gotham-Book",
                  "fontSize": 14,
                },
                Object {
                  "fontSize": 9,
                },
              ]
            }
          >
            OK
          </Text>
        </TouchableOpacity>
      </RowNB>
      <TicketLineControls
        active={0}
        isDiscount={true}
      />
    </TouchableOpacity>
  </TouchableOpacity>
</Modal>
`;
